plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java

```diff
@@ -84,7 +84,11 @@ public boolean parseDomainXML(String domXML) {
                     String authUserName = getAttrValue("auth", "username", disk);
                     String poolUuid = getAttrValue("secret", "uuid", disk);
                     String host = getAttrValue("host", "name", disk);
-                    int port = Integer.parseInt(getAttrValue("host", "port", disk));
+                    int port = 0;
+                    String _xmlPort = getAttrValue("host", "port", disk);
+                    if ( ! _xmlPort.isEmpty()) {
```

Comment:
Should we do both null and empty check?

Reply:
I would do that just for the sake of security

Reply:
there is a lot of npe danger in `getAttrValue()`, none of teh parameters are checked. (meaning yes please)

Reply:
I did an investigation of the dom code used in `getAttrValue()` and it returns an empty string if the element `"host"` does not have an attribute named `"port"`, so _xmlPort will not be null.

Reply:
1. @div8cn can you please change the variable name from `_xmlPort` to `xmlPort`?

According to [Java conventions](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html), 

> while it's technically legal to begin your variable's name with "_", this practice is discouraged

2. Regarding the String validation. Is it possible to have a `tagNode` of length 0?

`getAttrValue()` has this piece of code:
```
        NodeList tagNode = eElement.getElementsByTagName(tag);
        if (tagNode.getLength() == 0) {
            return null;
        }
```

I would recommend, just for safety, to use something as `org.apache.commons.lang.StringUtils`; it provides String validations such as `StringUtils.isNotBlank(Str)`.

`isNotBlank` returns true only if not null, not empty, and not blank.

```
String xmlPort = getAttrValue("host", "port", disk);
if (StringUtils.isNotBlank(xmlPort)) {
     ....
}
```